### PR TITLE
[InstCombine] Fold xored one-complemented operand comparisons

### LINKED ISSUE #69803: clang is suboptimal for `(~a ^ b) < ~a`
Consider the following two functions:
```cpp
int foo(int a, int b)
{
   return (~a ^ b) < ~a;
}

int bar(int a, int b)
{
   return (a ^ b) > a;
}
```

Clang vs GCC:
https://godbolt.org/z/rxK1nr794

Alive2:
https://alive2.llvm.org/ce/z/aUhKRe

```llvm
define i32 @src(i32 %0, i32 %1) {
2:
  %3 = xor i32 %0, 4294967295
  %4 = xor i32 %3, %1
  %5 = icmp slt i32 %4, %3
  %6 = zext i1 %5 to i32
  ret i32 %6
}
=>
define i32 @tgt(i32 %0, i32 %1) {
2:
  %3 = xor i32 %1, %0
  %4 = icmp sgt i32 %3, %0
  %5 = zext i1 %4 to i32
  ret i32 %5
}
Transformation seems to be correct!
```

### PR DESCRIPTION
**Pull request by elhewaty**
- [InstCombine] Add test coverage for comparisons of operands including one-complemented oparands(NFC).
- [InstCombine] Fold xored one-complemented operand comparisons.
Alive2: https://alive2.llvm.org/ce/z/PZMJeB
Fixes #69803.

### CODE REVIEW

**Comment by elhewaty** [comment]
@dtcxzyw please have a look.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:6935]: Please use `ICmpInst::getSwappedPredicate`.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7157]:
```suggestion
      return new ICmpInst(I.getSwappedPredicate(), Builder.CreateXor(A, B), A);
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7147]: Please provide alive2 proofs (Just attach the link to the PR description).

**Comment by elhewaty** [comment]
(~A ^ B) s< ~A --> (A ^ B) s> A
https://alive2.llvm.org/ce/z/kptBzu

(~A ^ B) s> ~A --> (A ^ B) s< A
https://alive2.llvm.org/ce/z/3_86kB

(~A ^ B) s<= ~A --> (A ^ B) s>= A
https://alive2.llvm.org/ce/z/qS7atw

(~A ^ B) s>= ~A --> (A ^ B) s<= A
https://alive2.llvm.org/ce/z/qS7atw

(~A ^ B) u< ~A --> (A ^ B) u< A
https://alive2.llvm.org/ce/z/4Z5Mpi

(~A ^ B) u> ~A --> (A ^ B) u< A
https://alive2.llvm.org/ce/z/mpY1Vb

(~A ^ B) u<= ~A --> (A ^ B) u>= A
https://alive2.llvm.org/ce/z/udZM2U

(~A ^ B) u>= ~A --> (A ^ B) <= A
https://alive2.llvm.org/ce/z/22u3Zm

File: llvm/test/Transforms/InstCombine/icmp-of-xor-x.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/icmp-of-xor-x.ll:7]: `zext` is unnecessary.
```
define i1 @test_slt_xor(i32 %0, i32 %1) {
  %3 = xor i32 %0, -1
  %4 = xor i32 %3, %1
  %5 = icmp slt i32 %4, %3
  ret i1 %5
}
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7156]: 1) You are overwriting `A` with the `m_Not` match. If you want to match the same `A` from the `xor` you would use `m_Specific(A)`.

2) You don't actually need `~A ^ B pred ~A`. You can do `~A ^ B pred ~C`:
https://alive2.llvm.org/ce/z/m8-MhQ

So the match on `Op1` should be setting a new value.

3) The match on `xor` should be commutative i.e it doesn't matter if its `~A ^ B` or `A ^ ~B`. so the match should be: `m_c_Xor(m_Not(A), B)`

4) This should work if `Op1` is the `m_Not` and `Op0` is the `m_Xor` so I would do a commutative match along the lines of:
`if(match(&I, m_c_ICmp(Pred, m_c_Xor(m_Not(m_Value(A)), m_Value(B)), m_Not(m_Value(C)))`.

File: llvm/test/Transforms/InstCombine/icmp-of-xor-x.ll
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/icmp-of-xor-x.ll:116]: Can you swap the operand order in some of the tests. I.e all your tests of of the form:
`~A ^ B pred A`. Can you:
1) Add tests where its form: `A pred Xor`
2) Add tests where the xor is of form `A ^ ~B`
3) Can you add tests for `~A ^ B pred ~C`

You don't need to add new tests for all these cases, just flip around operands in some of your current tests.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (elhewaty) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7156]: the `~A ^ B pred ~C` is already handled.

**Comment by elhewaty** [comment]
Alive2: https://alive2.llvm.org/ce/z/9DoJpD

File: llvm/test/Transforms/InstCombine/icmp-of-xor-x.ll
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/icmp-of-xor-x.ll:19]: This tests currrently has no diff from your patch.
`(xor (not A), B)` is canonicalized to `(not (xor A, B))` if `(not A)` has one use.

To make your tests meaningful you need to add something along the lines of:
```
declare void @use.i8(i8)
...
  call void @use.i8(i8 %xor)
```

at the end.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7130]: Please update comments.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7140]: This transform also works with `eq/ne`.
Alive2: https://alive2.llvm.org/ce/z/FEz2U8

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (elhewaty) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7140]: @I didn't write them because they are handled.
Aliv2: https://godbolt.org/z/qaTh55fsa

Comment (elhewaty) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7140]: I will remove the code that handles them.

**Comment by elhewaty** [comment]
@dtcxzyw why did the build fail?

**Comment by dtcxzyw** [comment]
> @dtcxzyw why did the build fail?

You should add `Value *C` to shadow the definition `const APInt *C` in the outer scope.

**Comment by goldsteinn** [comment]
> > @dtcxzyw why did the build fail?
> 
> You should add `Value *C` to shadow the definition `const APInt *C` in the outer scope.

Or instead rename variables to X/Y/Z as shadow variables can cause confusion and in llvm `C` usually means constant which isn't the case here.

**Comment by elhewaty** [comment]
@dtcxzyw @goldsteinn please take a look again.
Thanks.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7141]:
```suggestion
    if (match(&I, m_c_ICmp(Pred, m_c_Xor(m_Not(m_Value(X)), m_Value(Y)),
                           m_Not(m_Value(Z)))) && (I.getOperand(0)->hasOneUse() || I.getOperand(1)->hasOneUse()))
```
Please add a test that both operands are multi-use.
```
; Negative tests
define i1 @test_xor1_nofold_multi_use(i8 %x, i8 %y, i8 %z) {
  %xor = xor i8 %x, -1
  call void @use.i8(i8 %xor)
  %xor2 = xor i8 %xor, %y
  %nz = xor i8 %z, -1
  call void @use.i8(i8 %nz)
  %r = icmp slt i8 %xor2, %nz
  ret i1 %r
}
```

**Comment by elhewaty** [comment]
@dtcxzyw can you help me, what is wrong here?

**Comment by nikic** [comment]
@elhewaty This is due to an incorrect rebase around the processUMulZExtIdiom() call. Can you please rebase another time over https://github.com/llvm/llvm-project/commit/ea99df2e84d5f1da7f04bfa9f13844c44e0f1f78, which should remove all the diffs unrelated to your change?

File: llvm/test/Transforms/InstCombine/icmp-of-xor-x.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/icmp-of-xor-x.ll:279]: The extra use needs to be on %xor2 rather than %xor, otherwise the fold will still happen, so it's not a negative test.

File: llvm/test/Transforms/InstCombine/icmp-of-xor-x.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/icmp-of-xor-x.ll:178]:
```suggestion
define i1 @test_slt_xor(i32 %x, i32 %y) {
; CHECK-LABEL: @test_slt_xor(
; CHECK-NEXT:    [[TMP3:%.*]] = xor i32 [[TMP0:%.*]], [[TMP1:%.*]]
; CHECK-NEXT:    [[TMP4:%.*]] = icmp sgt i32 [[TMP3]], [[TMP0]]
; CHECK-NEXT:    ret i1 [[TMP4]]
;
  %xor = xor i32 %x, -1
  %xor2 = xor i32 %xor, %y
  %r = icmp slt i32 %xor2, %xor
  ret i1 %r
}
```
Looks like the tests starting from here ended up using unnamed values.

**Comment by elhewaty** [comment]
@dtcxzyw updated.

**Review by dtcxzyw — APPROVED** [review]
LGTM with the value name fix in `test_slt_xor` and `test_sle_xor`.

**Comment by elhewaty** [comment]
@dtcxzyw I used git clang-format but the check failed

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Comment by dtcxzyw** [comment]
> @dtcxzyw I used git clang-format but the check failed

Emm, it looks like an internal failure of the buildbot.

**Comment by elhewaty** [comment]
So, What is the future of the pull?
Will it land?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7044]:
```suggestion
    //           (~X ^ Y) u>= ~Z --> (X ^ Y) u<= Z,
    //           (~X ^ Y) == ~Z  --> (X ^ Y) == Z,
    // and       (~X ^ Y) != ~Z  --> (X ^ Y) != Z,
```

**Comment by elhewaty** [comment]
@dtcxzyw Please take a look.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

**Comment by elhewaty** [comment]
@nikic @goldsteinn @arsenm

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**
